### ISSUE TITLE
Ingesting of combined VCF (cVCF) is not supported

### ISSUE THREAD

**Issue by Hoeze** [TileDB-Inc/TileDB-VCF#50]
Hi, I just tried TileDB-VCF with a cVCF containing ~ 900 samples.
Unfortunately, I get the error message: 
```
Error registering samples; a file has more than 1 sample. Ingestion from cVCF is not supported.
```

I do not want to split a cVCF of this size into separate files. 
Is there an alternative?

**Comment by stavrospapadopoulos** [comment]
Unfortunately not currently, but we do understand the value of such a feature. Thanks for filing this, we will try to work on this soon.

**Comment by Hoeze** [comment]
@stavrospapadopoulos Would it be possible to have an ingestion API for non-VCF data?
I.e. write a dataframe or similar to the tiledb-vcf.
This way, we could develop custom ingestion algorithms.

**Comment by stavrospapadopoulos** [comment]
@Hoeze could you please provide an example? [Core TileDB](https://github.com/TileDB-Inc/TileDB) can store any dataframe very efficiently. Also see [related docs](https://docs.tiledb.com/main/use-cases/dataframes). 

cVCF support is special, since we need to be able to attach "axes labels" to the 2D matrix. We are working on this and hoping to publish it soon.

**Comment by Hoeze** [comment]
@stavrospapadopoulos In principle, I would expect to have something similar to:
```python
df = pd.DataFrame(..., columns=["contigName", "start", "alternative_sequence", "attributes", "samples_array"])
ds = tiledbvcf.TileDBVCFDataset('my-dataset', mode='w')
ds.write(df)
```
Might be improvable from the API point :)

**Comment by stavrospapadopoulos** [comment]
I'd like to first provide some clarifications around TileDB-VCF and how this differs from TileDB core, before suggesting a solution for this.

TileDB-VCF models a gVCF dataset (consisting of a collection of single-sample VCF files), as a special 2D "interval" array. It is an interval array in the sense that, at an abstract level, every entry of each row is a range interval, instead of a single point/cell. Therefore, every subarray query is essentially handled as "stabbing" query, i.e., instead of looking just for inclusion of cells, it is looking for intersections of the range intervals. TileDB-VCF does a couple of tricks and reduces this problem to a traditional 2D sparse array, and uses the TileDB core as a backend on that.

If you do _not_ need to store range intervals and you just have a traditional dataframe (e.g., a pandas/Spark/R dataframe), then you do not need to use TileDB-VCF. Instead, you can use directly the TileDB core, as a sparse array is essentially a generalization of a dataframe (since you can select any subset of columns as dimensions, effectively enforcing a primary index that allows fast multi-dimensional range search on your dataframe). The docs I pointed above contain more information. 

You can then read into a pandas dataframe simply as:
```
with tiledb.SparseArray(array_name, 'r', ctx=tiledb.Ctx(config)) as A:
    data = A[:]                                              # that returns a dictionary of numpy arrays
    print (pandas.DataFrame(data))          # works also with any attribute subselection
```

In other words, TileDB core probably provides all the functionality you need for arbitrary dataframe storage/access, whereas TileDB-VCF is suitable for "interval" arrays. 

cc-ing @ihnorton in case you need more information on how to ingest CSVs via pandas or read from TileDB core into pandas dataframes

**Comment by Hoeze** [comment]
Ah, I think I was unclear with my suggestion:
With "non-VCF" data I mean variant data which is not stored in a "\*.vcf" file but in any alternative format. For example, data stored in "BGEN" or "parquet" format, or even data which is calculated on the fly by some script.
In other words, it would be nice to have the ingestion algorithm accept in-memory data via an API instead of "*.vcf" file paths.

----
You explained that the basic addition of TileDB-VCF are range queries. As this is a more general problem which does not only apply to VCF files, would it make sense to have range queries in TileDB-core?

**Comment by stavrospapadopoulos** [comment]
> would it make sense to have range queries in TileDB-core?

Absolutely! It is already in our backlog.

> it would be nice to have the ingestion algorithm accept in-memory data via an API instead of "*.vcf" file paths

Yes, this makes a lot of sense. We just need different ingestors for the various formats (Parquet being perhaps the easiest). It is our vision to consolidate all those formats under our more general array model. We'll just need to prioritize based on demand, as it is indeed a very ambitious vision :). Please do suggest what is more pressing for you.

**Comment by Hoeze** [comment]
Are there any updates on the cVCF reader?
I do have a VCF file with 200,000 individuals that I would not like to split up into single VCF's :smile:

**Comment by stavrospapadopoulos** [comment]
No chance to get a hold of the original single-sample VCF files I assume, right? :) 

Even if we implement a cVCF ingestor, I am afraid we will ingest way too much redundancy into the TileDB arrays. In any case, we do have a cVCF ingestor in our roadmap, so we'll certainly keep you posted if we have any updates.

**Comment by Hoeze** [comment]
> No chance to get a hold of the original single-sample VCF files I assume, right? :)

Yep, unfortunately it's private data from UK Biobank.
Nevertheless, the file structure is very simple:
```
##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##GLnexusVersion=v1.2.6-0-g4d057dc
##GLnexusConfigName=DeepVariantWES
##GLnexusConfigCRC32C=4105299981
##GLnexusConfig={unifier_config: {drop_filtered: false, min_allele_copy_number: 1, min_AQ1: 35, min_AQ2: 20, min_GQ: 20, max_alleles_per_site: 32, monoallelic_sites_for_lost_alleles: true, preference: common}, genotyper_config: {revise_genotypes: true, min_assumed_allele_frequency: 9.99999975e-05, required_dp: 0, allow_partial_data: true, allele_dp_format: AD, ref_dp_format: MIN
##INFO=<ID=AF,Number=A,Type=Float,Description="Allele Frequency estimate for each alternate allele">
##INFO=<ID=AQ,Number=A,Type=Integer,Description="Allele Quality score reflecting evidence for each alternate allele (Phred scale)">
##INFO=<ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes">
##INFO=<ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes">
##FILTER=<ID=MONOALLELIC,Description="Site represents one ALT allele in a region with multiple variants that could not be unified into non-overlapping multi-allelic sites">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=RNC,Number=2,Type=Character,Description="Reason for No Call in GT: . = n/a, M = Missing data, P = Partial data, I = gVCF input site is non-called, D = insufficient Depth of coverage, - = unrepresentable overlapping deletion, L = Lost/unrepresentable allele (other than deletion), U = multiple Unphased variants present, O = multiple Overlapping variants present, 1 = s
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)">
##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
##FORMAT=<ID=PL,Number=G,Type=Integer,Description="Phred-scaled genotype Likelihoods">
##bcftools_viewVersion=1.2+htslib-1.2.1
##bcftools_viewCommand=view -
[...]
##contig=<ID=HLA-DRB1*07:01:01:01,length=16110>
[...]
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  [list of 200k samples]
chr10   46961   chr10_46961_G_A G       A       39      .       AF=7e-06;AQ=39  GT:DP:AD:GQ:PL:RNC [genotype fields]
```
The only speciality is the high number of chromosomes (3366) and the high number of samples (200,000).

> Even if we implement a cVCF ingestor, I am afraid we will ingest way too much redundancy into the TileDB arrays. In any case, we do have a cVCF ingestor in our roadmap, so we'll certainly keep you posted if we have any updates.

Which kind of redundancy?

---
Maybe this is of interest for you:
- Our in-house genotype array is a sparse 2D array with dimensions `(variant_key, sample)`. 
- We only store `(GT, GQ, DP)` attributes in this array.
- In addition, we do have a SQL database that contains all previously seen variants together with different types of annotations. 
We use this table to filter variants of interest and then query the TileDB array by a corresponding list of `variant_key`'s (the poor-man's join ^^).

**Comment by stavrospapadopoulos** [comment]
One of the problems with cVCF ingestion is that we need to get rid of the inflated cVCF entries for the samples that did not have those entries in their corresponding single-sample VCF files. In any case though, a cVCF ingestor has been requested frequently, so we'll try to implement it soon.

**Comment by stavrospapadopoulos** [comment]
We just implemented exported to cVCF form TileDB-VCF. Ingesting from cVCF is next on our list.

**Comment by gpertea** [comment]
As an interim  solution, just found that `bcftools +split` can be used to split a cVCF into individual vcf.gz files which can then be ingested  using `tiledbvcf store` - it worked well, did not take as long as I feared.

**Comment by gspowley** [comment]
Closing this issue since it has been inactive. Please reopen if there is still an issue.

**Comment by andreaswallberg** [comment]
I would like to ask for this issue and feature to be re-opened.

Combined VCFs are certainly the standard format for variation data in non-model organisms and having to split datasets before import is impractical.

**Comment by leipzig** [comment]
Hi Andreas,
Actually we now have methods to ingest multisample VCFs in our cloud product. We'll have that documented soon.

The krill VCFs unfortunately will have to be split lengthwise (with https://github.com/mlin/vcf_line_splitter for instance)
